packages/codegen/src/codegen.ts

```diff
@@ -0,0 +1,131 @@
+import { codegen } from '@graphql-codegen/core';
+import { preset } from '@graphql-codegen/near-operation-file-preset';
+import * as typedDocumentNodePlugin from '@graphql-codegen/typed-document-node';
+import * as typescriptOperationsPlugin from '@graphql-codegen/typescript-operations';
+import { printExecutableGraphQLDocument } from '@graphql-tools/documents';
+import { readFile, writeFile } from 'fs/promises';
+import { globby } from 'globby';
+import type { DocumentNode } from 'graphql';
+import { parse } from 'graphql';
+import * as path from 'node:path';
+import { defaultHash } from './default-hash.js';
+import { generateDocumentFiles } from './generate-document-files.js';
+import { generateSchemaTypes } from './generate-schema-types.js';
+import type { CodegenArgs, OutputFile } from './types.js';
+import { changeExtension } from './utils.js';
+
+export async function athenaCodegen({
+  schemaName,
+  documents,
+  baseDir,
+  extension,
+  hash,
+}: CodegenArgs): Promise<void> {
+  const outputFiles: Array<OutputFile> = [];
+  const persistedQueries: Record<string, string> = {};
+  const schemaPath = path.join(baseDir, schemaName);
+  // Read schema
+  const rawSchema = await readFile(schemaPath, 'utf-8');
+  const parsedSchema = parse(rawSchema);
+  const hashFn = hash || defaultHash;
```

Comment:
We should figure out how to allow the CLI to customize the hash function. Either as an option `--hash-function-path` or something. It might be good enough to just allow `--additional-cache-keys` though...

Reply:
@rwjblue i think the actual extension point needs to be something closer to `onExecutibleDocument` so that we're not making hard assumptions about things like the format of the persisted query file or even that they'll always be written to the same place (e.g. thinking about incremental rebuilds not having to know about all queries).

Maybe something like

```bash
codegen --plugins=@foo/bar
```

where `import default from '@foo/bar' returns something like

```ts
{
  onExecutibleDocumentNode(documentNode: DocumentNode, srcPath: string): void;
  // hook to signal that there is no more processing
  beforeExit(): Promise<void>;
}
``

Reply:
this is a good callout, but I think it should be a follow-on task for the time being